config/src/lib.rs

```diff
@@ -180,49 +301,111 @@ mod tests {
+            if value == None {
+                panic!("Argument was not specified!");
+            }
+            let value = value.expect("Shouldn't happen");
+            if value.len() != 1 {
+                panic!("Not a single value: {}", value.len());
+            }
+            String::from(&value[0].clone())
+        }
+
+        // Helper function to check whether the argument was provided on the command line
+        pub fn argument_was_provided(&self, key: &ConfigOption) -> bool {
+            if let Some(v) = self
+                .values
+                .get(key)
+                .expect("Fatal error: key not present in HashMap, but should have been!")
+            {
+                true
+            } else {
+                false
+            }
+        }
     }
 
-    impl ConfigDescription for TestConfig {
-        fn get_config(&self) -> Configuration {
+    // Implementation to return description of this config that is needed by ConfigBuilder
+    impl Configurable for TestConfig {
+        fn get_config_description() -> Configuration {
             Configuration {
                 name: "Test Tool",
                 version: "0.1",
                 about: "blabla",
-                options: vec![
+                options: [
                     TestConfig::TEST_PARAM,
                     TestConfig::TEST_PARAM2,
                     TestConfig::TEST_SWITCH,
-                ],
+                    TestConfig::TEST_MULTIPLE,
+                ]
+                .iter()
+                .cloned()
+                .collect(),
+            }
+        }
+
+        // Very simple implementation used for testing purposes only
+        // Simply store the HashMap
+        fn parse_values(parsed_values: HashMap<ConfigOption, Option<Vec<String>>>) -> Self {
+            TestConfig {
+                values: parsed_values,
             }
         }
     }
 
     #[test]
     fn parse_single_param() {
-        let config = TestConfig {};
         let command_line_args: Vec<OsString> = vec![
             OsString::from("filename"),
             OsString::from("--testparam"),
             OsString::from("param1"),
         ];
-        let matcher = get_matcher(&config, "Test", command_line_args)
-            .expect("unexpected error occurred when parsing parameters");
+        let config: TestConfig = ConfigBuilder::build(command_line_args, "test").expect("test");
 
-        assert!(matcher.is_present(TestConfig::TEST_PARAM.name));
+        // Check that absent parameters are reported correctly
         assert_eq!(
-            matcher.value_of(TestConfig::TEST_PARAM.name).unwrap(),
-            "param1"
+            config.argument_was_provided(&TestConfig::TEST_SWITCH),
+            false
         );
         assert_eq!(
-            matcher.value_of(TestConfig::TEST_PARAM2.name).unwrap(),
-            TestConfig::TEST_PARAM2.default
-        )
+            config.argument_was_provided(&TestConfig::TEST_PARAM2),
+            false
```

Comment:
You removed the check whether the default value is used. I think that'd be a useful test to have.

Reply:
Added those checks back in.